### DISCUSSION TITLE
Running hammerDB on windows client

### DISCUSSION THREAD

**Discussion by channeshcr** [TPC-Council/HammerDB#648] [Q&A] [answered]
I am running load test using hammerDB for the below setup, wanted to know how running hammerDB on windows/linux node present in the same subnet as database with different system resources affect the results? 

**Database:** oracle exadata (19c database version)
Database has 2 Linux nodes in a vm cluter each with 10 cores and 256gb ram. 
I have installed hammerDB on 2 windows nodes, which are in the same subnet as the database. One with just 1 core and other with 8 cores CPU. 
I got the below results when I ran the load test against 500 warehouses with 50 vusers, keying and thinking disabled and Use All warehouses enabled. 
On windows node with 1 core cpu
**System achieved 197857 NOPM from 417526 Oracle TPM**
On windows node with 8 core cpu
**System achieved 221912 NOPM from 465135 Oracle TPM**

I understood that the schema creation depends on the node where we run the hammerDB from, as it took more than an hour on 1 core machine and less than 30 mins on 8 core machine.

If we are to run the load tests on a windows/linux node where the database is not installed and these nodes are in the same network as database, what are the resource considerations we should keep in mind ? 
Also, should we consider running load tests on the database node itself ? 
How would these scenarios affect the results ?

**Comment by sm-shaw** [comment]
The advice is the same as given here https://github.com/TPC-Council/HammerDB/discussions/649 - if you have 10 cores use fewer virtual users to start with to find the maximum performance. 

The HammerDB workload is light so will not place a lot of load on your client system. 

 When running HammerDB on the same system as the database it is a trade off between the CPU HammerDB uses running locally against the network overhead running remotely.  Oracle networking is efficient, and the database can effectively use all cores so the approach in our experience for maximum performance is to use a remote client. Of course to save time you can install HammerDB locally and do the schema build from there and then run the test remotely. 

Don't forget HammerDB can generate an AWR report for you and you can view the metrics in real time (drag and drop the window out to expand the view) so you have all the features you need to diagnose performance. 

![ch7-12](https://github.com/TPC-Council/HammerDB/assets/38044085/17ffa91f-a532-42ba-a0ef-93062d09e15f)

**Reply by channeshcr** [reply]
Thanks @sm-shaw for the detailed explanation. Cleared my doubt about the usage of remote client. 
I am still exploring the options, i have these basic questions about warehouse and virtual users. This is probably a common question you get all the time :)

I have read on the hammerDB doc that 4-5 warehouses per vuser would be a minimum value to ensure an even distribution of vusers to warehouses. And one warehouse is roughly about 100 MB. 
How many warehouses one should consider while building the schema ?  For a setup like mine with 10 cores and 256 GB RAM, is 100 warehouses ideal for my setup ? Or I can have any number(assuming i have enough storage) of warehouses and do the load testing starting with 1 vuser and then build up in increments using autopilots ?

**Reply by sm-shaw** [reply, marked as answer]
Yes, and just for clarity for anyone following up on this thread, remote client is best from experience for Oracle, with open source databases that cannot fully utilize all the CPU on the database server, local is often best as HammerDB is not using CPU that the database could. 

For the number of warehouses there is an explanation in the introductory video here around 30 mins https://www.hammerdb.com/about.html.

But, basically in the default usage, the virtual users pick a home warehouse at random where most of the activity will take place and therefore the 4-5 warehouses per vuser is to allow enough of a distribution to ensure that in most cases a vuser has its own warehouse. So in this default usage then for 10 cores then yes 100 warehouses would be fine and yes as in the video 100, 500 or 1000 should not make a great deal of difference as as each virtual user will work on a home warehouse regardless of how many there are in total. 

Note that use all warehouses will divide all of the warehouses between vusers and then randomly change the home warehouse at each iteration and therefore if this is enabled more warehouses means more phyiscal I/O is generated. Here there will be a difference between 100 and 1000 - this is what the feature is meant for. 

The first thing I would always recommend is to [calibrate the CPU ](https://www.hammerdb.com/docs/ch04s01.html) to make sure there are no obvious throttles to performance - having power saving enabled, reducing CPU frequency is a very common one.

Then, always start with 1 Vuser, next 2, 4 8 etc and see the difference between them. With Oracle you can generate an AWR report and observe the top timed wait events to diagnose performance.

**Comment by channeshcr** [comment]
Thanks again @sm-shaw
